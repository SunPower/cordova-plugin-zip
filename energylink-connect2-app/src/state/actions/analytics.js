import { namedAction } from 'shared/redux-utils'

const analyticsAction = namedAction('Analytics')
const mixpanelActions = namedAction('Mixpanel')

export const BEGIN_INSTALL = analyticsAction('BEGIN_INSTALL')
export const COMMISSIONING_FINISH = analyticsAction('FINISH COMMISSIONING')
export const COMMISSION_SUCCESS = analyticsAction('COMMISSION_SUCCESS')
export const CONFIG_START = analyticsAction('Configure start')
export const CONNECT_PVS_CAMERA = analyticsAction('CONNECT_PVS_CAMERA')
export const CONNECT_PVS_MANUALLY = analyticsAction('CONNECT_PVS_MANUALLY')
export const FINISH_PLT_SETUP = analyticsAction('Finish PLT setup')
export const MIXPANEL_EVENT_ERROR = analyticsAction('MIXPANEL_EVENT_ERROR')
export const MIXPANEL_EVENT_QUEUED = mixpanelActions('Event queued')
export const RESET_PVS_INTERNET_TRACKING = analyticsAction(
  'RESET_PVS_INTERNET_TRACKING'
)
export const SCANNING_START = analyticsAction('Scanning start')
export const SET_AC_DEVICES = analyticsAction('Set AC modules')
export const CLAIM_DEVICES_MIXPANEL_EVENT = analyticsAction(
  'CLAIM DEVICES MIXPANEL EVENT'
)
export const START_BULK_SETTINGS_TIMER = analyticsAction('Start bulk settings')
export const START_PLT_SETUP = analyticsAction('Start PLT setup')
