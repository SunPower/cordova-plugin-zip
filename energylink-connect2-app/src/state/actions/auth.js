import { createAction } from 'redux-act'
import {
  pipe,
  equals,
  head,
  last,
  prop,
  merge,
  omit,
  identity,
  converge,
  useWith
} from 'ramda'
import { getFile } from './fileDownloader'
import authClient from 'shared/auth/sdk'
import { httpGet } from 'shared/fetch'

export const LOGIN_INIT = createAction('LOGIN_INIT')
export const LOGIN_SUCCESS = createAction('LOGIN_SUCCESS')
export const LOGIN_ERROR = createAction('LOGIN_ERROR')
export const LOGOUT = createAction('LOGOUT')

const ROLES = {
  PARTNER: 'partner',
  PARTNER_PRO: 'partner_pro'
}

export const requestLogin = () => {
  return (dispatch, state) => {
    try {
      let state = authClient.generateRandomValue()
      dispatch(LOGIN_INIT({ state }))
      authClient.authorizeOAuth(state)
    } catch (err) {
      dispatch(LOGIN_ERROR({ message: err }))
    }
  }
}

export const handleLoginFromPing = URL => {
  return (dispatch, getState) => {
    const { user } = getState()
    const hashes = authClient.parseURL(URL)

    if (hashes.error && hashes.error_description) {
      const error = `${hashes.error}: ${hashes.error_description}`
      dispatch(LOGIN_ERROR({ message: error }))
      return
    }

    if (hashes.state && !equals(hashes.state, user.auth.state)) {
      dispatch(LOGIN_ERROR({ message: 'STATE_MISMATCH' }))
      return
    }

    if (hashes.code) {
      authClient
        .getAccessTokenOAuth(hashes.code)
        .then(token => {
          dispatch(handleUserProfile(token))
        })
        .catch(error => {
          dispatch(LOGIN_ERROR({ message: 'FETCH_ACCESS_TOKEN' }))
        })
    }
  }
}

const discardUserPropsFromAPIResponse = pipe(
  prop('data'),
  omit(['userId', 'displayName', 'email'])
)

const merge_EDP_PING_Response = useWith(merge, [
  discardUserPropsFromAPIResponse,
  identity
])

const buildUser = converge(merge_EDP_PING_Response, [head, last])

export const handleUserProfile = (tokenInfo = {}) => {
  return dispatch => {
    try {
      const { access_token } = tokenInfo
      Promise.all([
        httpGet('/auth/user', null, access_token),
        authClient.getUserInfoOAuth(access_token)
      ])
        .then(buildUser)
        .then(user => {
          const payload = { data: user, auth: tokenInfo }
          const ug = user.userGroup.toLowerCase()
          if (ug !== ROLES.PARTNER && ug !== ROLES.PARTNER_PRO) {
            dispatch(LOGIN_ERROR({ message: 'INVALID_ROLE' }))
          } else {
            dispatch(LOGIN_SUCCESS(payload))
            //dispatch(fetchInventory())
            dispatch(getFile())
          }
        })
        .catch(error => {
          console.error(error)
          const details = prop('details', error)

          if (!details) {
            dispatch(LOGIN_ERROR({ message: 'HTTP_ERROR' }))
            return
          }

          const firstError = head(details)
          const firstErrorCode = prop('code', firstError)

          if (equals(firstErrorCode, 'INVALID_VALUE')) {
            const firstErrorMsg = prop('message', firstError)
            const ERROR = firstErrorMsg.includes('Access token expired')
              ? 'ACCESS_TOKEN_EXPIRED'
              : firstErrorMsg
            dispatch(LOGIN_ERROR({ message: ERROR }))
          } else {
            const description = prop('error_description', error)
            dispatch(LOGIN_ERROR({ message: description }))
          }
        })
    } catch (err) {
      dispatch(LOGIN_ERROR(err))
    }
  }
}

export const verifyToken = access_token => {
  return dispatch => {
    authClient
      .verifyTokenOAuth(access_token)
      .then(response => {
        if (!response.uniqueId) {
          dispatch(LOGIN_ERROR({ message: 'ACCESS_TOKEN_EXPIRED' }))
          dispatch(LOGOUT())
          return
        }
        dispatch(VALIDATE_SESSION_SUCCESS())
      })
      .catch(error => {
        dispatch(LOGIN_ERROR({ message: 'ACCESS_TOKEN_EXPIRED' }))
        dispatch(LOGOUT())
        dispatch(VALIDATE_SESSION_ERROR())
      })
  }
}

export const VALIDATE_SESSION_INIT = createAction('VALIDATE_SESSION_INIT')
export const VALIDATE_SESSION_SUCCESS = createAction('VALIDATE_SESSION_SUCCESS')
export const VALIDATE_SESSION_ERROR = createAction('VALIDATE_SESSION_ERROR')

export const validateSession = () => {
  return (dispatch, getState) => {
    const { user } = getState()
    if (user.auth.access_token) {
      dispatch(VALIDATE_SESSION_INIT())
      dispatch(verifyToken(user.auth.access_token))
    }
  }
}
