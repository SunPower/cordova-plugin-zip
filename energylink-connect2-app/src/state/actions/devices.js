import { createAction } from 'redux-act'

export const DISCOVER_INIT = createAction('discover devices init')
export const DISCOVER_UPDATE = createAction('discover devices update')
export const DISCOVER_COMPLETE = createAction('discover devices complete')
export const DISCOVER_ERROR = createAction('discover devices error')
export const PUSH_CANDIDATES_INIT = createAction('PUSH_CANDIDATES_INIT')
export const PUSH_CANDIDATES_SUCCESS = createAction('PUSH_CANDIDATES_SUCCESS')
export const PUSH_CANDIDATES_ERROR = createAction('PUSH_CANDIDATES_ERROR')
export const FETCH_CANDIDATES_INIT = createAction('FETCH_CANDIDATES_INIT')
export const FETCH_CANDIDATES_UPDATE = createAction('FETCH_CANDIDATES_UPDATE')
export const FETCH_CANDIDATES_ERROR = createAction('FETCH_CANDIDATES_ERROR')
export const FETCH_CANDIDATES_COMPLETE = createAction(
  'FETCH_CANDIDATES_COMPLETE'
)
export const CLAIM_DEVICES_INIT = createAction('CLAIM_DEVICES_INIT')
export const CLAIM_DEVICES_SUCCESS = createAction('CLAIM_DEVICES_SUCCESS')
export const CLAIM_DEVICES_ERROR = createAction('CLAIM_DEVICES_ERROR')
export const RESET_DISCOVERY = createAction('RESET_DISCOVERY')
