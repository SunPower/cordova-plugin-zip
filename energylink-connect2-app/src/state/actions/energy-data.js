import { createAction } from 'redux-act'

export const FETCH_LTE_DATA_INIT = createAction('FETCH_LTE_DATA_INIT')
export const FETCH_LTE_DATA_SUCCESS = createAction('FETCH_LTE_DATA_SUCCESS')
export const FETCH_LTE_DATA_ERROR = createAction('FETCH_LTE_DATA_ERROR')

export const FETCH_CURRENT_POWER_INIT = createAction('FETCH_CURRENT_POWER_INIT')
export const FETCH_CURRENT_POWER_SUCCESS = createAction(
  'FETCH_CURRENT_POWER_SUCCESS'
)
export const FETCH_CURRENT_POWER_ERROR = createAction(
  'FETCH_CURRENT_POWER_ERROR'
)

export const GET_ENERGY_DATA_INIT = createAction('GET_ENERGY_DATA_INIT')
export const GET_ENERGY_DATA_SUCCESS = createAction('GET_ENERGY_DATA_SUCCESS')
export const GET_ENERGY_DATA_ERROR = createAction('GET_ENERGY_DATA_ERROR')

export const LIVE_ENERGY_DATA_NOTIFICATION = createAction(
  'LIVE_ENERGY_DATA_NOTIFICATION'
)
export const LIVE_ENERGY_DATA_DAILY = createAction('LIVE_ENERGY_DATA_DAILY')
export const LIVE_ENERGY_DATA_CONNECTION = createAction(
  'LIVE_ENERGY_DATA_CONNECTION'
)
export const LIVE_ENERGY_DATA_ERROR = createAction('LIVE_ENERGY_DATA_ERROR')

export const GET_POWER_DATA_INIT = createAction('GET_POWER_DATA_INIT')
export const GET_POWER_DATA_SUCCESS = createAction('GET_POWER_DATA_SUCCESS')
export const GET_POWER_DATA_ERROR = createAction('GET_POWER_DATA_ERROR')

export const INTERVALS = {
  FIVE_MINUTE: 'five_minute',
  QUARTER_HOUR: 'quarter_hour',
  HOUR: 'hour',
  DAY: 'day',
  MONTH: 'month',
  YEAR: 'year',
  LIVE: 'live'
}

// Minimum amount of time that has to pass before fetching more data
export const DELTAS = {
  [INTERVALS.HOUR]: 1000 * 60 * 5 /* 5 minutes */,
  [INTERVALS.DAY]: 1000 * 60 * 30 /* 30 minutes */,
  [INTERVALS.MONTH]: 1000 * 60 * 60 * 12 /* 12 hours */,
  [INTERVALS.YEAR]: 1000 * 60 * 60 * 24 * 7 /* 1 week */
}

export const ENERGY_DATA_START_POLLING = createAction(
  'ENERGY_DATA_START_POLLING'
)
export const ENERGY_DATA_POLLING = createAction('ENERGY_DATA_POLLING')
export const ENERGY_DATA_STOP_POLLING = createAction('ENERGY_DATA_STOP_POLLING')

export const POWER_DATA_START_POLLING = createAction('POWER_DATA_START_POLLING')
export const POWER_DATA_POLLING = createAction('POWER_DATA_POLLING')
export const POWER_DATA_STOP_POLLING = createAction('POWER_DATA_STOP_POLLING')

export const CURRENT_POWER_DATA_START_POLLING = createAction(
  'CURRENT_POWER_DATA_START_POLLING'
)
export const CURRENT_POWER_DATA_STOP_POLLING = createAction(
  'CURRENT_POWER_DATA_STOP_POLLING'
)

export const NORMALIZE_ENERGY_DATA_INIT = createAction(
  'NORMALIZE_ENERGY_DATA_INIT'
)
export const NORMALIZE_ENERGY_DATA_SUCCESS = createAction(
  'NORMALIZE_ENERGY_DATA_SUCCESS'
)

export const NORMALIZE_POWER_DATA_INIT = createAction(
  'NORMALIZE_POWER_DATA_INIT'
)
export const NORMALIZE_POWER_DATA_SUCCESS = createAction(
  'NORMALIZE_POWER_DATA_SUCCESS'
)

export const LTE_DATA_POLL_START = createAction('LTE_DATA_POLL_START')
export const LTE_DATA_POLL_STOP = createAction('LTE_DATA_POLL_STOP')
