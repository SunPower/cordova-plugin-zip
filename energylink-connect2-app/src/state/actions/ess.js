import { createAction } from 'redux-act'

export const DOWNLOAD_META_INIT = createAction('DOWNLOAD_META_INIT')
export const DOWNLOAD_META_SUCCESS = createAction('DOWNLOAD_META_SUCCESS')
export const DOWNLOAD_META_ERROR = createAction('DOWNLOAD_META_ERROR')

export const DOWNLOAD_OS_INIT = createAction('DOWNLOAD_OS_INIT')
export const DOWNLOAD_OS_SUCCESS = createAction('DOWNLOAD_OS_SUCCESS')
export const DOWNLOAD_OS_PROGRESS = createAction('DOWNLOAD_OS_PROGRESS')
export const DOWNLOAD_OS_ERROR = createAction('DOWNLOAD_OS_ERROR')
