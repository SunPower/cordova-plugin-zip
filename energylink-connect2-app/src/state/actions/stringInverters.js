import { createAction } from 'redux-act'

export const SAVE_STRING_INVERTERS = createAction('SAVE_STRING_INVERTERS')
