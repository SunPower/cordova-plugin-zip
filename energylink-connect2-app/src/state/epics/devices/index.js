import { claimDevicesEpic } from './claimDevices'
import { scanDevicesEpic } from './discoverDevices'
import { fetchDeviceListEpic } from './fetchDeviceList'
import { fetchModelsEpic } from './fetchModels'
import { fetchCandidatesEpic } from './pollCandidates'
import { pollClaimingEpic } from './pollClaiming'
import { pushCandidatesEpic } from './pushCandidates'

export default [
  pushCandidatesEpic,
  fetchCandidatesEpic,
  scanDevicesEpic,
  claimDevicesEpic,
  pollClaimingEpic,
  fetchModelsEpic,
  fetchDeviceListEpic
]
