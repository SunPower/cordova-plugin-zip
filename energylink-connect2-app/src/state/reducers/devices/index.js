import { createReducer } from 'redux-act'
import { length } from 'ramda'
import {
  DISCOVER_COMPLETE,
  DISCOVER_INIT,
  DISCOVER_UPDATE,
  DISCOVER_ERROR,
  FETCH_CANDIDATES_INIT,
  FETCH_CANDIDATES_UPDATE,
  FETCH_CANDIDATES_COMPLETE,
  FETCH_CANDIDATES_ERROR,
  CLAIM_DEVICES_INIT,
  CLAIM_DEVICES_SUCCESS,
  CLAIM_DEVICES_ERROR,
  RESET_DISCOVERY
} from 'state/actions/devices'

const initialState = {
  isFetching: false,
  found: {},
  progress: {},
  error: '',
  isFetchingCandidates: false,
  candidates: [],
  allCandidatesFound: false,
  discoveryComplete: false,
  claimingDevices: false,
  claimedDevices: false
}

const parseCompleteDevices = devices => {
  const returnValue = {}
  devices.forEach(device => {
    const deviceType = device.DEVICE_TYPE.toLowerCase()
    const propertyExists = !!returnValue[deviceType]
    if (propertyExists)
      returnValue[deviceType] = [...returnValue[deviceType], device]
    else returnValue[deviceType] = [device]
  })
  return returnValue
}

export default createReducer(
  {
    [DISCOVER_INIT]: state => ({ ...state, isFetching: true }),
    [DISCOVER_UPDATE]: (state, payload) => {
      return {
        ...state,
        found: payload
          ? parseCompleteDevices(payload.devices.devices)
          : state.found,
        progress: payload ? payload.progress : state.progress
      }
    },
    [DISCOVER_COMPLETE]: (state, payload) => {
      return {
        ...state,
        isFetching: false,
        found: payload
          ? parseCompleteDevices(payload.devices.devices)
          : state.found,
        progress: payload ? payload.progress : state.progress,
        discoveryComplete: true
      }
    },
    [DISCOVER_ERROR]: (state, payload) => {
      return {
        ...state,
        isFetching: false,
        progress: {},
        error: payload
      }
    },
    [FETCH_CANDIDATES_INIT]: state => {
      return {
        ...state,
        isFetchingCandidates: true
      }
    },
    [FETCH_CANDIDATES_UPDATE]: (state, payload) => {
      return {
        ...state,
        isFetchingCandidates: true,
        candidates: length(payload) !== 0 ? payload : state.candidates
      }
    },
    [FETCH_CANDIDATES_COMPLETE]: state => {
      return {
        ...state,
        isFetchingCandidates: false,
        allCandidatesFound: true
      }
    },
    [FETCH_CANDIDATES_ERROR]: (state, payload) => {
      return {
        ...state,
        isFetchingCandidates: false,
        candidates: [],
        error: payload
      }
    },
    [CLAIM_DEVICES_INIT]: state => {
      return {
        ...state,
        claimingDevices: true
      }
    },
    [CLAIM_DEVICES_SUCCESS]: state => {
      return {
        ...state,
        claimingDevices: false,
        claimedDevices: true
      }
    },
    [CLAIM_DEVICES_ERROR]: (state, payload) => {
      return {
        ...state,
        claimingDevices: false,
        error: payload
      }
    },
    [RESET_DISCOVERY]: () => {
      return {
        ...initialState
      }
    }
  },
  initialState
)
