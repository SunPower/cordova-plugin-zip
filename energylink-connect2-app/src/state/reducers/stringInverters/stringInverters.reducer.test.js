import stringInverterReducer, { initialState } from './index'

import { FETCH_MODELS_SUCCESS } from 'state/actions/devices'
import { SAVE_STRING_INVERTERS } from 'state/actions/stringInverters'

describe('The stringInverterReducer', function() {
  const reducerTest = reducerTester(stringInverterReducer)
  describe('The initial state', function() {
    it('should not change', function() {
      expect(initialState).toMatchSnapshot()
    })
  })
  it('should get a new state with SAVE_STRING_INVERTERS', function() {
    const newDevices = [
      { serialNumber: '123', moduleCount: '2' },
      { serialNumber: '234', moduleCount: '2' }
    ]
    reducerTest(initialState, SAVE_STRING_INVERTERS(newDevices), {
      ...initialState,
      newDevices
    })
  })
  it('should get a new state with FETCH_MODELS_SUCCESS', function() {
    const responseCall = {
      A: ['SPR-240E-WHT-U ACPV', 'SPR-245NE-WHT-U-240 ACPV'],
      B: ['SPR-E20-245-B-AC', 'SPR-X20-250-BLK AC'],
      C: [
        'SPR-E19-320-C-AC',
        'SPR-E20-327-C-AC',
        'SPR-X21-335-BLK-C-AC',
        'SPR-X21-335-C-AC',
        'SPR-X21-345-C-AC',
        'SPR-X22-360-C-AC'
      ],
      D: [
        'SPR-E20-327-D-AC',
        'SPR-X19-315-D-AC',
        'SPR-X20-327-BLK-D-AC',
        'SPR-X20-327-D-AC',
        'SPR-X21-335-BLK-D-AC',
        'SPR-X21-335-D-AC',
        'SPR-X21-345-D-AC',
        'SPR-X21-350-BLK-D-AC',
        'SPR-X22-360-D-AC',
        'SPR-X22-370-D-AC',
        'SPR-240E-WHT-D AR'
      ],
      E: [
        'SPR-E19-320-E-AC',
        'SPR-E20-327-E-AC',
        'SPR-X20-327-BLK-E-AC',
        'SPR-X20-327-E-AC',
        'SPR-X21-335-BLK-E-AC',
        'SPR-X21-335-E-AC',
        'SPR-X21-345-E-AC',
        'SPR-X21-350-BLK-E-AC',
        'SPR-X22-360-E-AC',
        'SPR-X22-370-E-AC',
        'SPR-X21-355-E-AC'
      ],
      G: [
        'SPR-A390-G-AC',
        'SPR-A400-G-AC',
        'SPR-A410-G-AC',
        'SPR-A415-G-AC',
        'SPR-A420-G-AC',
        'SPR-A390-BLK-G-AC',
        'SPR-A400-BLK-G-AC'
      ],
      H: [
        'SPR-A390-H-AC',
        'SPR-A400-H-AC',
        'SPR-A410-H-AC',
        'SPR-A415-H-AC',
        'SPR-A420-H-AC',
        'SPR-A390-BLK-H-AC',
        'SPR-A400-BLK-H-AC'
      ],
      'V1.0': ['SPR-225E-BLK-U ACPV'],
      //these are not string inverter modules, they're model strings that represent solar panels that are attached to string inverters
      stringInverters: initialState.models
    }
    reducerTest(initialState, FETCH_MODELS_SUCCESS(responseCall), {
      ...initialState,
      models: responseCall.stringInverters
    })
  })
})
