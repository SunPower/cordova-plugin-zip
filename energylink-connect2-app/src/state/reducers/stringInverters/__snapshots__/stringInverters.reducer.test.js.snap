// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stringInverterReducer The initial state should not change 1`] = `
Object {
  "models": Array [
    "SPR-A390-BLK",
    "SPR-A400-BLK",
    "SER-220P",
    "SER-228P",
    "SER-228P-T10",
    "SER-235P",
    "SPR-200-BLK-U",
    "SPR-205-BLK-U",
    "SPR-210-BLK-U",
    "SPR-210-WHT-U",
    "SPR-215-BLK-U",
    "SPR-215-WHT-U",
    "SPR-217-WHT-U C",
    "SPR-220-BLK-U N",
    "SPR-220-WHT-U",
    "SPR-225E-BLK-D",
    "SPR-225E-WHT-D",
    "SPR-230E-WHT-D",
    "SPR-230NE-BLK-D",
    "SPR-230-T10",
    "SPR-230-WHT-U",
    "SPR-238E-WHT-D",
    "SPR-290-WHT-D",
    "SPR-305E-WHT-D",
    "SPR-305-T10",
    "SPR-305-T5",
    "SPR-310-WHT-D",
    "SPR-318-T5",
    "SPR-327NE-WHT-D",
    "SPR-400E-WHT-D",
    "SPR-425E-WHT-D",
    "SPR-A430-COM",
    "SPR-A430-COM-MLSD",
    "SPR-A440-COM",
    "SPR-A440-COM-MLSD",
    "SPR-A450-COM",
    "SPR-A450-COM-MLSD",
    "SPR-E19-310-COM",
    "SPR-E19-320",
    "SPR-E19-410-COM",
    "SPR-E20-245",
    "SPR-E20-320-COM",
    "SPR-E20-320-COM Helix",
    "SPR-E20-320-COM MLSD",
    "SPR-E20-320-COM MLSD Helix",
    "SPR-E20-327",
    "SPR-E20-327-COM",
    "SPR-E20-327-COM Helix",
    "SPR-E20-327-COM MLSD",
    "SPR-E20-327-COM MLSD Helix",
    "SPR-E20-330-COM",
    "SPR-E20-330-COM Helix",
    "SPR-E20-330-COM MLSD",
    "SPR-E20-330-COM MLSD Helix",
    "SPR-E20-435-COM",
    "SPR-P15-315-COM",
    "SPR-P17-330-COM",
    "SPR-P17-340-COM",
    "SPR-P17-340-COM H",
    "SPR-P17-345-COM",
    "SPR-P17-350-COM",
    "SPR-P17-350-COM H",
    "SPR-P17-355-COM",
    "SPR-P17-355-COM H",
    "SPR-P17-360-COM",
    "SPR-P17-360-COM H",
    "SPR-P19-380-COM",
    "SPR-P19-380-COM H",
    "SPR-P19-380-COM MLSD",
    "SPR-P19-380-COM MLSD H",
    "SPR-P19-385-COM",
    "SPR-P19-385-COM H",
    "SPR-P19-385-COM MLSD",
    "SPR-P19-385-COM MLSD H",
    "SPR-P19-390-COM",
    "SPR-P19-390-COM H",
    "SPR-P19-390-COM MLSD",
    "SPR-P19-390-COM MLSD H",
    "SPR-P19-395-COM",
    "SPR-P19-395-COM H",
    "SPR-P19-395-COM MLSD",
    "SPR-P19-395-COM MLSD H",
    "SPR-P19-400-COM",
    "SPR-P19-400-COM H",
    "SPR-P19-400-COM MLSD",
    "SPR-P19-400-COM MLSD H",
    "SPR-X20-327",
    "SPR-X20-327-BLK",
    "SPR-X21-335",
    "SPR-X21-335-BLK",
    "SPR-X21-345",
    "SPR-X21-345-COM",
    "SPR-X21-345-COM MLSD",
    "SPR-X21-345-COM MLSD Helix",
    "SPR-X21-350-BLK",
    "SPR-X21-460-COM",
    "SPR-X21-470-COM",
    "SPR-X22-360",
    "SPR-X22-360-COM",
    "SPR-X22-360-COM MLSD",
    "SPR-X22-360-COM MLSD Helix",
    "SPR-X22-370",
    "PL-PLT-63-BLK-U",
    "SPR-76RE-BLK-U",
  ],
  "newDevices": Array [],
}
`;
